Blog: Unlock Programmatic Control by Disabling Default Providers (#1241)
pulumi/pulumi-hugo: themes/default/content/blog/disabledefaultproviders/index.md

---
title: "Unlock Programmatic Control by Disabling Default Providers"
date: 2022-03-18

> meta_desc: Introducing the new config parameter disable-default-providers.

Author: If you an existing card in mind or the time to make a card that would be awesome. I'm going to look through older blog posts and see if anything jumps out at me.

> meta_desc: Introducing the new config parameter disable-default-providers.

Author: I borrowed the card from https://www.pulumi.com/blog/take-control-with-pac/, but I also thought of using the card for https://www.pulumi.com/blog/retainondelete/, but replace the stamp content with "disable default providers". What do you think works best?

> meta_desc: Introducing the new config parameter disable-default-providers.

Reviewer: I'm so sorry I missed your reply; I was heads down on a demo. How does this one work for you?
![meta (5)](https://user-images.githubusercontent.com/1538692/160639706-15ad13a9-ed87-429a-af3e-ca4fac8b4db0.png)

authors: [ ian-wahbe ]
tags: [ features ]
---

> As of 3.23.0, users can disable the default provider with Pulumi. So what does this mean for you? If you’ve been using Pulumi for a bit, you’ll have encountered provider resources, which are how we abstract the global state of a cloud provider. All resources have an associated provider. If no provider is supplied in the user’s code, a default provider is created to serve the resource. On the other hand, explicit providers defined by the user allow programmatic and dynamic control of how a resource deploys into a cloud. A Pulumi resource can be instructed to use an explicit provider by setting the provider resource option or by inheriting the provider from the resource's parent resource.

Reviewer: ```suggestion
As of 3.23.0, users can disable the default provider with Pulumi. So what does this mean for you? If you’ve been using Pulumi for a bit, you’ll have encountered [provider resources]({{< relref "/docs/intro/concepts/resources/providers" >}}), which are how we abstract the global state of a cloud provider. All resources have an associated provider. If no provider is supplied in the user’s code, a [default provider]({{< relref "/docs/intro/concepts/resources/providers#default-provider-configuration" >}}) is created to serve the resource. On the other hand, explicit providers defined by the user allow programmatic and dynamic control of how a resource deploys into a cloud. A Pulumi [resource]({{< relref "/docs/intro/concepts/resources" >}}) can be instructed to use an explicit provider by setting the [provider resource option]({{< relref "/docs/intro/concepts/resources/options/provider" >}}) or by inheriting the provider from the resource's [parent resource]({{< relref "/docs/intro/concepts/resources/options/parent" >}}).
```

Reviewer: I don't know why GitHub isn't picking up what I changed, but I added links in from the google doc.

	return err
}

bucket, err := s3.NewBucket(ctx, "my-bucket", &s3.BucketArgs{
	Website: s3.BucketWebsiteArgs{
	IndexDocument: pulumi.StringPtr("index.html"),
	},
}, pulumi.Provider(west))
if err != nil {
	return err
}

_, err = s3.NewBucketObject(ctx, "index.html", &s3.BucketObjectArgs{
	Acl:         s3.CannedAclPublicRead,
	ContentType: pulumi.StringPtr("text/html"),
	Bucket:      bucket,
	Source:      pulumi.NewFileAsset("index.html"),
}, pulumi.Parent(bucket))
if err != nil {
	return err
}
```

{{% /choosable %}}

{{% choosable language csharp %}}

```csharp
var west = new Aws.Provider("provider", new Aws.ProviderArgs {
        Region = "us-east-1",
});

var bucket = new S3.Bucket("my-bucket", new S3.BucketArgs {
        Website = new Aws.S3.Inputs.BucketWebsiteArgs {
            IndexDocument = "index.html"
        },
    },
    new CustomResourceOptions { Provider = west },
);

new S3.BucketObject("index.html", new S3.BucketObjectArgs {
        Acl = "public-read",
        ContentType = "text/html",
        Bucket = bucket,
        Source = FileAsset("index.html"),
     },
    new CustomResourceOptions { Parent = bucket },
);
```

{{% /choosable %}}

{{< /chooser >}}

> Default providers help keep simple things simple but make other actions impossible. As demonstrated in the above example, AWS providers, whether they are explicit or default, control the deployment region. That means multi-region deployments necessitate explicit providers. Likewise, deploying to a newly created Kubernetes cluster requires explicit providers since Kubernetes providers describe their cluster. There are many scenarios when explicit providers are mandatory for a correct deployment. When it is important that only explicitly configured providers are used, default providers lead to unpredictable deployments.

Reviewer 2: >When it is important that only explicitly configured providers are used, default providers lead to unpredictable deployments.

"Unpredictable" sounds a bit scary. Is it accurate? I can see how mixing explicit and default providers within the same program could lead to confusion or surprise, but not so much unpredictability. (Unless I'm misunderstanding of course.) Would it still be accurate to put it this way instead?

```suggestion
Default providers help keep simple things simple but make other actions impossible. As demonstrated in the above example, AWS providers, whether they are explicit or default, control the deployment region. That means multi-region deployments necessitate explicit providers. Likewise, deploying to a newly created Kubernetes cluster requires explicit providers since Kubernetes providers describe their cluster. There are many scenarios when explicit providers are mandatory for a correct deployment. When it is important that only explicitly configured providers are used, default providers can lead to unexpected results.
```

```
Error: Invoke: Default provider for 'aws' disabled. 'aws:index/getRegions:getRegions' must use an explicit provider.
```

To prevent this category of problem, Pulumi now offers the ability to disable default providers on a per-stack basis.  

Disabling default providers has been a popular community request. To meet that request, we’ve added the ability to disable default providers in the 3.23.0 release. The Pulumi config variable `pulumi:disable-default-providers` represents the list of packages whose default providers are disabled. Attempting to invoke a disabled default provider will fail, raising an error as in the previous example.

## How to disable default providers

Disabling a default provider involves adding the relevant package name to the config list `pulumi:disable-default-providers`. For example, disabling the default provider for Kubernetes and AWS would look like this snippet in the configuration file:

```yaml
pulumi:disable-default-providers:
- aws
- kubernetes
```

If you want to enforce that no resource should use a default provider, you would add the following snippet to the configuration file:

```yaml
pulumi:disable-default-providers: [“*”]
```

> Now that you can disable the default provider, you don’t have to worry about all of the possible unexpected consequences accidentally relying on your system configuration. We can’t wait to find out what you’ll build next! If you want to have an impact on our roadmap, you can go to our public roadmap and vote on issues with the rest of the community. Let us know what you think!

Reviewer 2: Just adding a couple of potentially helpful links:

```suggestion
Now that you can [disable the default provider]({{< relref "/docs/intro/concepts/resources/providers#disabling-default-providers" >}}), you don’t have to worry about all of the possible unexpected consequences accidentally relying on your system configuration.

We can’t wait to find out what you’ll build next! If you want to have an impact on our roadmap, you can go to our [public roadmap](https://github.com/orgs/pulumi/projects/44) and vote on issues with the rest of the community. Let us know what you think!
```
